atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1138 @@
+    /// * `message_id` - Unique identifier for the gossipsub message
+    /// * `propagation_source` - The peer that forwarded this message
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if message processing succeeds, or an error if:
+    /// - Message deserialization fails
+    /// - Message validation fails
+    /// - Reporting validation result fails
+    /// - Storing public URL fails (for clients)
+    ///
+    /// # Errors
+    ///
+    /// This function can return the following errors:
+    /// * `UsageMetricsSerializeError` - If CBOR deserialization fails
+    /// * `GossipsubMessageRebroadcastError` - If reporting validation result fails
+    /// * `StateManagerError` - If storing public URL fails (for clients)
+    ///
+    /// # Security Considerations
+    ///
+    /// - Messages from self are ignored to prevent message loops
+    /// - Messages are validated before processing
+    /// - Only clients store public URLs to prevent unnecessary data storage
+    /// - Uses CBOR for efficient binary serialization
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// let message_data = // ... received from gossipsub ...;
+    /// let message_id = // ... message identifier ...;
+    /// let propagation_source = // ... peer ID ...;
+    ///
+    /// match node.handle_gossipsub_message(&message_data, &message_id, &propagation_source).await {
+    ///     Ok(()) => println!("Message processed successfully"),
+    ///     Err(e) => println!("Failed to process message: {}", e),
+    /// }
+    /// ```
+    ///
+    /// # Message Validation
+    ///
+    /// Messages are validated by:
+    /// 1. Checking the URL format and timestamp freshness
+    /// 2. Verifying the cryptographic signature
+    /// 3. Confirming the node's ownership of its small ID
+    ///
+    /// Invalid messages are rejected and not propagated further in the network.
+    #[instrument(level = "debug", skip_all)]
+    pub async fn handle_gossipsub_message(
+        &mut self,
+        message_data: &[u8],
+        message_id: &gossipsub::MessageId,
+        propagation_source: &PeerId,
+    ) -> Result<(), AtomaP2pNodeError> {
+        debug!(
+            target = "atoma-p2p",
+            event = "gossipsub_message",
+            message_id = %message_id,
+            propagation_source = %propagation_source,
+            "Received gossipsub message"
+        );
+        if propagation_source == self.swarm.local_peer_id() {
+            debug!(
+                target = "atoma-p2p",
+                event = "gossipsub_message_from_self",
+                "Gossipsub message from self"
+            );
+            // Do not re-publish the node's own message, just return `Ok(())
+            return Ok(());
+        }
+        let gossip_message: GossipMessage = ciborium::from_reader(message_data)?;
+        match gossip_message {
+            GossipMessage::UsageMetrics(usage_metrics) => {
```

Comment:
Do we intend to gossip other types of messages for the time being? Otherwise I don't think `UsageMetrics` needs to be an enum here, we can just use the type as done [here](https://github.com/atoma-network/atoma-node/pull/286/files#diff-f558264be9dc74d18752e874e9306be2942281ddfc84762ae59c17ae15b74d64R730)

Reply:
Yes, I do believe we want to add other gossip message types